plugins/OnlineSales.Plugin.SendGrid/Tasks/SyncSuppressionsTask.cs

```diff
@@ -46,57 +48,56 @@ public SyncSuppressionsTask(TaskStatusService taskStatusService, PgDbContext dbC
 
     public override async Task<bool> Execute(TaskExecutionLog currentJob)
     {
-        var bounces = await GetLatestSuppressionsByType<BlockOrBounceDto>("bounces");
-
-        foreach (var bounce in bounces)
-        {
-            await contactService.Unsubscribe(bounce.Email, $"{bounce.Status} - {bounce.Reason}", "SendGrid_bounces", bounce.CreatedAt, null);
-        }
-
-        var blocks = await GetLatestSuppressionsByType<BlockOrBounceDto>("blocks");
-
-        foreach (var block in blocks)
-        {
-            await contactService.Unsubscribe(block.Email, $"{block.Status} - {block.Reason}", "SendGrid_blocks", block.CreatedAt, null);
-        }
-
-        var spamReports = await GetLatestSuppressionsByType<SpamReportDto>("spam_reports");
+        await Unsubscribe<BlockOrBounceDto>("bounces");
+        await Unsubscribe<BlockOrBounceDto>("blocks");
+        await Unsubscribe<SpamReportDto>("spam_reports");
+        await Unsubscribe<SuppressionDto>("unsubscribes");       
+        return true;
+    }
 
-        foreach (var spamReport in spamReports)
-        {
-            await contactService.Unsubscribe(spamReport.Email, "Reported As Spam", "SendGrid_spam_reports", spamReport.CreatedAt, spamReport.Ip);
-        }
+    private string GetSecondaryApiKeyName(int i)
+    {
+        return $"SecondaryApiKey{i}";
+    }
 
-        var unsubscribes = await GetLatestSuppressionsByType<SuppressionDto>("unsubscribes");
+    private async Task Unsubscribe<T>(string suppressionType)
+        where T : SuppressionDto
+    {
+        var primaryKeyData = await GetLatestSuppressionsByType<T>(primaryApiKeyName, SendGridPlugin.Configuration.SendGridApi.PrimaryApiKey, suppressionType);
 
-        foreach (var unsubscribe in unsubscribes)
+        foreach (var pkd in primaryKeyData)
```

Comment:
Why do not you make a single list of all API keys and do not combine the logic inside two foreach loops into a single piece of code?